File: Python_sol/Linked_List/Leetcode_19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        counter = 1
        temp = head
        while temp:
            temp = temp.next
            counter += 1
        
        
        temp = head
        prev = temp
        for pos in range(1, counter):
            if pos == counter - n:
                if temp == prev:
                    head = head.next
                    break
                
                temp = temp.next
                prev.next = temp
                break
            prev = temp
            temp = temp.next
        return head

File: Javascript_sol/Linked_List/Leetcode_19.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let temp = head;
    let counter = 0;
    while (temp) {
        temp = temp.next;
        ++counter;
    };
    // Time complexity O(N?) Space complexity O(1)
    // The best solution would require two pointers
    if (counter != n) {
        temp = eval("head" + ".next".repeat(counter - n - 1));
    } else {
        head = head.next;
        return head;
    };
    temp.next = temp.next.next;
    
    return head;
};